monkey/monkey_island/cc/services/node.py

```diff
@@ -207,6 +220,43 @@ def insert_node(ip_address, domain_name=''):
             })
         return mongo.db.node.find_one({"_id": new_node_insert_result.inserted_id})
 
+    @staticmethod
+    def create_node_from_bootloader_telem(bootloader_telem: Dict, will_monkey_run: bool):
+        new_node_insert_result = mongo.db.node.insert_one(
+            {
+                "ip_addresses": bootloader_telem['ips'],
+                "domain_name": bootloader_telem['hostname'],
+                "will_monkey_run": will_monkey_run,
+                "exploited": False,
+                "creds": [],
+                "os":
+                    {
+                        "type": bootloader_telem['system'],
+                        "version": bootloader_telem['os_version']
+                    }
+            })
+        return mongo.db.node.find_one({"_id": new_node_insert_result.inserted_id})
+
+    @staticmethod
+    def get_or_create_node_from_bootloader_telem(bootloader_telem: Dict, will_monkey_run: bool) -> Dict:
+        if is_local_ips(bootloader_telem['ips']):
+            raise NodeCreationException("Bootloader ran on island, no need to create new node.")
+
+        new_node = mongo.db.node.find_one({"ip_addresses": {"$in": bootloader_telem['ips']}})
+        if new_node is None:
+            new_node = NodeService.create_node_from_bootloader_telem(bootloader_telem, will_monkey_run)
+            if bootloader_telem['tunnel']:
+                dst_node = NodeService.get_node_or_monkey_by_ip(bootloader_telem['tunnel'])
+            else:
+                dst_node = NodeService.get_monkey_island_node()
+            edge = EdgeService.get_or_create_edge(new_node['_id'], dst_node['id'])
+            mongo.db.edge.update({"_id": edge["_id"]},
+                                 {'$set': {'tunnel': bool(bootloader_telem['tunnel']),
+                                           'ip_address': bootloader_telem['ips'][0],
+                                           'group': NodeStates.get_by_keywords(['island']).value}},
+                                 upsert=False)
+        return new_node
+
```

Comment:
Any chance to cover this code with UTs? Or do you prefer to wait until we refactor `node` into `models`?

Reply:
`create_node_from_bootloader_telem` is just a mongo insert. `get_or_create_node_from_bootloader_telem` has a bit of logic, but it should change when we refactor map generation